torch/_inductor/codegen/cpp_prefix.h

```diff
@@ -96,6 +96,22 @@ template <typename T> void atomic_add(volatile T *addr, T offset) {
                                                std::memory_order_relaxed));
 }
 
+template <typename T> void atomic_fetch_add(volatile T *addr, T offset) {
+  static_assert(sizeof(std::atomic<T>) == sizeof(T),
+                "std::atomic issue");
+  std::atomic<T> *atomic_addr = (std::atomic<T> *)addr;
+  atomic_addr->fetch_add(offset, std::memory_order_relaxed);
+}
+
+// Since C++20 float is supported by fetch_add, but the performance may not
+// better than compare_exchange_weak, which can be checked by microbenchmark
+// inductor_cpu_atomic.py
+template <> inline void atomic_add<unsigned char>(volatile unsigned char *addr, unsigned char offset) { atomic_fetch_add(addr, offset); }
+template <> inline void atomic_add<signed char>(volatile signed char *addr, signed char offset) { atomic_fetch_add(addr, offset); }
+template <> inline void atomic_add<short>(volatile short *addr, short offset) { atomic_fetch_add(addr, offset); }
+template <> inline void atomic_add<int>(volatile int *addr, int offset) { atomic_fetch_add(addr, offset); }
+template <> inline void atomic_add<long>(volatile long *addr, long offset) { atomic_fetch_add(addr, offset); }
```

Comment:
I guess you can simplify this with SINFAE like below
```c++
template <class T>
inline typename std::enable_if<std::is_integral<T>::value>::type
atomic_add(volatile T *addr, T offset) {
  atomic_fetch_add(addr, offset);
}
```

Reply:
Yes, I did try it yesterday, but the ambiguous overload problem is annoying when T is float, unless I add the inverted template (i.e., typename std::enable_if<!std::is_integral<T>::value>::type) to the origin atomic_add function.
Do you know any better solution? Thanks.

Reply:
> Yes, I did try it yesterday, but the ambiguous overload problem is annoying when T is float, unless I add the inverted template (i.e., typename std::enable_if<!std::is_integral::value>::type) to the origin atomic_add function. Do you know any better solution? Thanks.

I think it is just fine to add an inverted condition on the floating point implementation.